test/e2e/apps/deployment.go

```diff
@@ -785,6 +788,127 @@ func testDeploymentsControllerRef(f *framework.Framework) {
+
+	firstRS, err := deploymentutil.GetNewReplicaSet(deployment, c.ExtensionsV1beta1())
+	Expect(err).NotTo(HaveOccurred())
+
+	// Update the deployment with a non-existent image so that the new replica set will be blocked.
+	framework.Logf("Updating deployment %q with a non-existent image", deploymentName)
+	deployment, err = framework.UpdateDeploymentWithRetries(c, ns, d.Name, func(update *extensions.Deployment) {
+		update.Spec.Template.Spec.Containers[0].Image = "nginx:404"
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	framework.Logf("Waiting for observed generation %d", deployment.Generation)
+	Expect(framework.WaitForObservedDeployment(c, ns, deploymentName, deployment.Generation)).NotTo(HaveOccurred())
+
+	framework.Logf("Verifying that deployment %q has minimum required number of available replicas", deploymentName)
+	minAvailableReplicas := deploymentutil.MinAvailable(deployment)
+	if deployment.Status.AvailableReplicas < minAvailableReplicas {
+		Expect(fmt.Errorf("observed %d available replicas, less than min required %d", deployment.Status.AvailableReplicas, minAvailableReplicas)).NotTo(HaveOccurred())
+	}
+
+	framework.Logf("Verifying that the replicasets of deployment %q have desired number of replicas", deploymentName)
+	firstRS, err = c.ExtensionsV1beta1().ReplicaSets(ns).Get(firstRS.Name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	secondRS, err := deploymentutil.GetNewReplicaSet(deployment, c.ExtensionsV1beta1())
+	Expect(err).NotTo(HaveOccurred())
+
+	err = framework.WaitForReplicaSetDesiredReplicas(c.ExtensionsV1beta1(), firstRS)
+	Expect(err).NotTo(HaveOccurred())
+	err = framework.WaitForReplicaSetDesiredReplicas(c.ExtensionsV1beta1(), secondRS)
+	Expect(err).NotTo(HaveOccurred())
+
+	framework.Logf("Verifying that first rollout's replicaset has minimum required number of available replicas")
+	if firstRS.Status.AvailableReplicas < minAvailableReplicas {
+		Expect(fmt.Errorf("observed %d available replicas, less than min required %d", firstRS.Status.AvailableReplicas, minAvailableReplicas)).NotTo(HaveOccurred())
+	}
+
+	availableReplicas := replicas - deployment.Spec.Strategy.RollingUpdate.MaxUnavailable.IntVal
+	framework.Logf("Waiting for first rollout's replicaset to have %d available replicas", availableReplicas)
+	err = framework.WaitForReplicaSetAvailableReplicas(c.ExtensionsV1beta1(), firstRS, availableReplicas)
+	Expect(err).NotTo(HaveOccurred())
+
+	framework.Logf("Verifying that second rollout's replicaset does not have any available replicas")
+	Expect(secondRS.Status.AvailableReplicas).Should(Equal(int32(0)))
+
+	newReplicas := int32(15)
+	framework.Logf("Scaling up the deployment %q from %d to %d", deploymentName, replicas, newReplicas)
+	deployment, err = framework.UpdateDeploymentWithRetries(c, ns, deployment.Name, func(update *extensions.Deployment) {
+		update.Spec.Replicas = &newReplicas
+	})
+	Expect(err).NotTo(HaveOccurred())
```

Comment:
We can use scale subresource to avoid conflict. But let's leave it for now and revisit it later.

Reply:
ok

Reply:
I meant the update conflict, which is why we update deployment with retries.